dbms/src/Storages/S3/FileCache.cpp

```diff
@@ -0,0 +1,563 @@
+#include <Storages/S3/FileCache.h>
+#include <Storages/S3/S3Common.h>
+#include <aws/s3/model/GetObjectRequest.h>
+
+#include <atomic>
+#include <chrono>
+#include <fstream>
+
+namespace ProfileEvents
+{
+extern const Event S3GetObject;
+extern const Event S3ReadBytes;
+extern const Event FileCacheHit;
+extern const Event FileCacheMiss;
+extern const Event FileCacheEvict;
+} // namespace ProfileEvents
+
+namespace DB::ErrorCodes
+{
+extern const int S3_ERROR;
+}
+
+namespace DB
+{
+
+using FileType = FileSegment::FileType;
+
+FileCache::FileCache(const String & cache_dir_, UInt64 cache_capacity_, UInt64 cache_level_)
+    : cache_dir(cache_dir_)
+    , cache_capacity(cache_capacity_)
+    , cache_level(cache_level_)
+    , cache_used(0)
+    , log(Logger::get("FileCache"))
+{
+    prepareDir(cache_dir);
+    restore();
+}
+
+RandomAccessFilePtr FileCache::getRandomAccessFile(const S3::S3FilenameView & s3_fname)
+{
+    auto file_seg = get(s3_fname);
+    if (file_seg == nullptr)
+    {
+        return nullptr;
+    }
+    // PosixRandomAccessFile should hold the `file_seg` shared_ptr to prevent cached file from evicted.
+    return std::make_shared<PosixRandomAccessFile>(file_seg->getLocalFileName(), /*flags*/ -1, /*read_limiter*/ nullptr, file_seg);
+}
+
+FileSegmentPtr FileCache::get(const S3::S3FilenameView & s3_fname)
+{
+    auto s3_key = s3_fname.toFullKey();
+    auto file_type = getFileType(s3_key);
+    auto & table = tables[static_cast<UInt64>(file_type)];
+
+    std::lock_guard lock(mtx);
+
+    auto f = table.get(s3_key);
+    if (f != nullptr)
+    {
+        f->setLastAccessTime(std::chrono::system_clock::now());
+        if (f->isReadyToRead())
+        {
+            ProfileEvents::increment(ProfileEvents::FileCacheHit);
+            return f;
+        }
+        else
+        {
+            ProfileEvents::increment(ProfileEvents::FileCacheMiss);
+            return nullptr;
+        }
+    }
+    ProfileEvents::increment(ProfileEvents::FileCacheMiss);
+
+    if (!canCache(file_type))
+    {
+        // Don't cache this file type or too many downloading task.
+        return nullptr;
+    }
+
+    // File not exists, try to download and cache it in backgroud.
+
+    // We don't know the exact size of a object/file, but we need reserve space to save the object/file.
+    // A certain amount of space is reserved for each file type.
+    auto estimzted_size = getEstimatedSizeOfFileType(file_type);
+    if (!reserveSpaceImpl(file_type, estimzted_size, /*try_evict*/ true))
+    {
+        LOG_DEBUG(log, "s3_key={} space not enough, skip cache", s3_key);
+        // Space not enough.
+        return nullptr;
+    }
+
+    auto file_seg = std::make_shared<FileSegment>(toLocalFilename(s3_key), FileSegment::Status::Empty, estimzted_size, file_type);
+    table.set(s3_key, file_seg);
+    bgDownload(s3_key, file_seg);
+
+    return nullptr;
+}
+
+// Remove `local_fname` from disk and remove parent directory if parent directory is empty.
```

Comment:
Will there be any concurrent issues? For example, while a thread is trying to remove parent dir and another thread is trying to put files in it.